# Module 2 Report  - Jobs
### Extracting SF Salaries Insights
<br>
Billie Kim
<br>
Data Source: https://www.kaggle.com/datasets/san-francisco/sf-employee-compensation

In [1]:
# Load SF Employee Compensation data
df <- read.csv("../mod_report2/data/employee-compensation.csv", stringsAsFactors=FALSE, na.strings="")

In [2]:
suppressWarnings(library(tidyverse))

Registered S3 methods overwritten by 'ggplot2':
  method         from 
  [.quosures     rlang
  c.quosures     rlang
  print.quosures rlang
Registered S3 method overwritten by 'rvest':
  method            from
  read_xml.response xml2
── Attaching packages ─────────────────────────────────────── tidyverse 1.2.1 ──
✔ ggplot2 3.1.1       ✔ purrr   0.3.2  
✔ tibble  2.1.1       ✔ dplyr   0.8.0.1
✔ tidyr   0.8.3       ✔ stringr 1.4.0  
✔ readr   1.3.1       ✔ forcats 0.4.0  
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()


## 1. Objective

## 2. Data Exploration

In [3]:
cat("Number of rows:", nrow(df))
cat("\nNumber of columns:", ncol(df))

Number of rows: 683277
Number of columns: 22

**Comment**: We are dealing with a fairly large dataset with 683,277 records and 22 columns.

In [4]:
head(df, n=10)

Organization.Group.Code Job.Family.Code Job.Code Year.Type Year
1  1                       1000            1021     Calendar  2013
2  1                       1000            1023     Calendar  2013
3  1                       1000            1031     Calendar  2013
4  1                       1000            1054     Calendar  2013
5  1                       1000            1062     Calendar  2013
6  1                       1000            1064     Calendar  2013
7  1                       1000            1070     Calendar  2013
8  1                       1000            1094     Calendar  2013
9  1                       1200            1232     Calendar  2013
10 1                       1600            1654     Calendar  2013
   Organization.Group Department.Code Department          Union.Code
1  Public Protection  ADP             ADP Adult Probation 21        
2  Public Protection  ADP             ADP Adult Probation 21        
3  Public Protection  ADP             ADP Adult Probation 21        
4  Public Protection  ADP             ADP Adult Probation 21        
5  Public Protection  ADP             ADP Adult Probation 21        
6  Public Protection  ADP             ADP Adult Probation 21        
7  Public Protection  ADP             ADP Adult Probation 21        
8  Public Protection  ADP             ADP Adult Probation 21        
9  Public Protection  ADP             ADP Adult Probation 21        
10 Public Protection  ADP             ADP Adult Probation 21        
   Union                                           ... Employee.Identifier
1  Prof & Tech Engineers - Miscellaneous, Local 21 ... 37730              
2  Prof & Tech Engineers - Miscellaneous, Local 21 ... 26574              
3  Prof & Tech Engineers - Miscellaneous, Local 21 ...  8148              
4  Prof & Tech Engineers - Miscellaneous, Local 21 ... 27436              
5  Prof & Tech Engineers - Miscellaneous, Local 21 ... 37730              
6  Prof & Tech Engineers - Miscellaneous, Local 21 ... 36223              
7  Prof & Tech Engineers - Miscellaneous, Local 21 ... 51426              
8  Prof & Tech Engineers - Miscellaneous, Local 21 ... 26574              
9  Prof & Tech Engineers - Miscellaneous, Local 21 ... 37983              
10 Prof & Tech Engineers - Miscellaneous, Local 21 ... 16858              
   Salaries  Overtime Other.Salaries Total.Salary Retirement Health.and.Dental
1   57534.65 0            0.00        57534.65    11253.16   11961.34         
2   57678.50 0            0.00        57678.50    10284.88    7176.80         
3   63532.93 0            0.00        63532.93    12495.83   12918.24         
4  101274.51 0        -7058.59        94215.92    19644.45   12918.24         
5    5084.00 0            0.00         5084.00     1083.90     956.91         
6   34456.00 0            0.00        34456.00     7346.00    3827.63         
7  139889.00 0            0.00       139889.00    26799.23   12918.24         
8   46572.00 0            0.00        46572.00     9929.16    5741.44         
9   84681.53 0            0.00        84681.53    16661.84   12918.24         
10  10611.01 0            0.00        10611.01     1879.20    1435.36         
   Other.Benefits Total.Benefits Total.Compensation
1  4532.63        27747.13        85281.78         
2  4755.14        22216.82        79895.32         
3  4702.71        30116.78        93649.71         
4  7458.76        40021.45       134237.37         
5   387.50         2428.31         7512.31         
6  2579.95        13753.58        48209.58         
7  9791.54        49509.01       189398.01         
8  3822.28        19492.88        66064.88         
9  6838.44        36418.52       121100.05         
10  827.09         4141.65        14752.66

**Comment:** We can't see all of the columns since there are too many so we will print out the column names next.

In [5]:
length(unique(df$Employee.Identifier))

[1] 309750

In [6]:
names(df)

[1] "Organization.Group.Code" "Job.Family.Code"        
 [3] "Job.Code"                "Year.Type"              
 [5] "Year"                    "Organization.Group"     
 [7] "Department.Code"         "Department"             
 [9] "Union.Code"              "Union"                  
[11] "Job.Family"              "Job"                    
[13] "Employee.Identifier"     "Salaries"               
[15] "Overtime"                "Other.Salaries"         
[17] "Total.Salary"            "Retirement"             
[19] "Health.and.Dental"       "Other.Benefits"         
[21] "Total.Benefits"          "Total.Compensation"

In [7]:
# str(df)
glimpse(df)

Observations: 683,277
Variables: 22
$ Organization.Group.Code <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, …
$ Job.Family.Code         <chr> "1000", "1000", "1000", "1000", "1000", "1000…
$ Job.Code                <chr> "1021", "1023", "1031", "1054", "1062", "1064…
$ Year.Type               <chr> "Calendar", "Calendar", "Calendar", "Calendar…
$ Year                    <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 201…
$ Organization.Group      <chr> "Public Protection", "Public Protection", "Pu…
$ Department.Code         <chr> "ADP", "ADP", "ADP", "ADP", "ADP", "ADP", "AD…
$ Department              <chr> "ADP Adult Probation", "ADP Adult Probation",…
$ Union.Code              <int> 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 2…
$ Union                   <chr> "Prof & Tech Engineers - Miscellaneous, Local…
$ Job.Family              <chr> "Information Systems", "Information Systems",…
$ Job                     <chr> "IS Administrator 1", "IS Administrator 3", "…
$ Employee.Ident

In [8]:
sapply(df, class)

Organization.Group.Code         Job.Family.Code                Job.Code 
              "integer"             "character"             "character" 
              Year.Type                    Year      Organization.Group 
            "character"               "integer"             "character" 
        Department.Code              Department              Union.Code 
            "character"             "character"               "integer" 
                  Union              Job.Family                     Job 
            "character"             "character"             "character" 
    Employee.Identifier                Salaries                Overtime 
              "integer"               "numeric"               "numeric" 
         Other.Salaries            Total.Salary              Retirement 
              "numeric"               "numeric"               "numeric" 
      Health.and.Dental          Other.Benefits          Total.Benefits 
              "numeric"               "numeric"               "numeric" 
     Total.Compensation 
              "numeric"

## 3. Subsets of Data

In [9]:
job.salaries <- df %>% 
    select(Job,Job.Family,Salaries,Total.Compensation) %>%
    filter(Salaries > 0)

head(job.salaries, n=5)

Job                           Job.Family          Salaries 
1 IS Administrator 1            Information Systems  57534.65
2 IS Administrator 3            Information Systems  57678.50
3 IS Trainer-Assistant          Information Systems  63532.93
4 IS Business Analyst-Principal Information Systems 101274.51
5 IS Programmer Analyst         Information Systems   5084.00
  Total.Compensation
1  85281.78         
2  79895.32         
3  93649.71         
4 134237.37         
5   7512.31

In [10]:
nrow(job.salaries)

[1] 671457

In [11]:
# job.salaries %>% 
#     arrange(desc(Salaries))

In [12]:
head(job.salaries %>%
    count(Job.Family, sort=T), n=10)

# job.salaries %>%
#     count(Job.Family, sort=T, wt=Salaries)

Job.Family                    n    
1  Nursing                       70127
2  Street Transit                55214
3  Police Services               41503
4  Journeyman Trade              39005
5  Human Services                34713
6  Public Service Aide           31305
7  Clerical, Secretarial & Steno 29146
8  Fire Services                 27619
9  Housekeeping & Laundry        21863
10 Management                    20884

### 3.1 Average Salaries, Compensation by Job Family

In [13]:
# avg.salaries <- df %>% 
#     group_by(Job.Family) %>%
#     summarize(count = n(), avg_salaries = mean(Salaries), avg_compensation = mean(Total.Compensation), percent = 100 * n()/sum(nrow(df))) %>%
#     arrange(desc(avg_salaries))

avg.salaries.df <- df %>% 
    group_by(Job.Family) %>%
    summarize(avg_salaries = mean(Salaries)) %>%
    arrange(desc(avg_salaries)) %>%
    ungroup

head(avg.salaries.df, n=10)

Job.Family                    avg_salaries
1  Administrative & Mgmt (Unrep) 237005.3    
2  Unassigned                    147929.4    
3  Construction Project Mgmt     126806.2    
4  Administrative-DPW/PUC        124543.1    
5  Management                    120669.6    
6  Medical & Dental              117536.1    
7  Legal & Court                 113306.6    
8  Administrative-Labor & Trades 110067.2    
9  Fire Services                 105072.4    
10 Port Operation                103672.1

### 3.2 Top 10 Highest Paying Jobs

In [14]:
top.paid.jobs <- df %>% 
    select(Job, Department, Total.Compensation) %>%
    arrange(desc(Total.Compensation))

In [15]:
head(top.paid.jobs, n=10)

Job                         Department            Total.Compensation
1  Police Officer 3            Police                807625.2          
2  Chief Investment Officer    Retirement Services   735942.8          
3  Chief Investment Officer    Retirement Services   724031.1          
4  Police Officer 3            Police                721649.9          
5  Chief Investment Officer    Retirement Services   707130.0          
6  Chief Investment Officer    Retirement Services   701272.4          
7  Chief Investment Officer    Retirement Services   685601.7          
8  Chief Investment Officer    Retirement Services   677935.5          
9  Dep Chf of Dept (Fire Dept) Fire Department       668896.5          
10 Chief Investment Officer    RET Retirement System 668412.4

In [16]:
top_10_jobs <- df %>%
  group_by(Job) %>%
  summarise(Total_Compensation = mean(Total.Compensation)) %>%
  arrange(desc(Total_Compensation)) %>%
  head(10)

print(top_10_jobs)

# A tibble: 10 x 2
   Job                            Total_Compensation
   <chr>                                       <dbl>
 1 Chief Investment Officer                  496297.
 2 Physician Administrator, DPH              488861.
 3 Managing Director                         451704.
 4 Sup Psychiatric Physician Spec            441229.
 5 Chief, Fire Department                    436281.
 6 Administrator, DPH                        418256.
 7 Director                                  400767.
 8 Assistant Chief of Police                 398627.
 9 Controller                                377979.
10 Gen Mgr, Public Trnsp Dept                377815.


### 3.3 Yearly Salary Trends

In [17]:
yearly.salary <- df %>%
    group_by(Year) %>%
    summarise(avg_salary = mean(Salaries), avg_overtime = mean(Overtime), avg_total_salary = mean(Total.Salary))

In [18]:
head(yearly.salary, n=8)

Year avg_salary avg_overtime avg_total_salary
1 2013 63801.78   4249.039     72023.35        
2 2014 62089.04   4305.350     70218.11        
3 2015 62884.41   4411.050     70920.06        
4 2016 65688.75   4928.004     74309.36        
5 2017 71180.22   5675.529     80290.33        
6 2018 76083.92   6490.329     85970.82        
7 2019 76222.49   7076.575     86788.81        
8 2020 82828.64   7306.643     93996.42

### 3.4 Union vs Non-Union 

In [19]:
union <- df %>%
    group_by(Union) %>%
    summarise(avg_salary = mean(Salaries), avg_total_benefits = mean(Total.Benefits), avg_total_compensation = mean(Total.Compensation)) %>%
    arrange(desc(avg_salary))

In [20]:
head(union)

Union                                    avg_salary avg_total_benefits
1 Unrepresented Contract Rte FBP           330199.8   94823.93          
2 Management Unrepresented Employees - MTA 305029.5   94299.66          
3 Indv. Employment Contract-MTA            264807.3   75609.68          
4 Municipal Exec Assoc, Police             258464.7   73991.33          
5 Physician/Dentists 11-AA, UAPD           239092.1   73227.59          
6 Executive Contract Employees             228060.2   74467.03          
  avg_total_compensation
1 425559.5              
2 399329.1              
3 353611.6              
4 353232.8              
5 324426.1              
6 328158.6

In [21]:
nrow(union)

[1] 130

### 3.5 Union Employee Count

In [22]:
count.union <- df %>%
    group_by(Union) %>%
    summarise(count = n()) %>%
    arrange(desc(count))

head(count.union, n=10)

Union                                              count 
1  SEIU, Local 1021, Misc                             115502
2  SEIU - Miscellaneous, Local 1021                   108284
3  Prof & Tech Engineers - Miscellaneous, Local 21     48814
4  Prof & Tech Eng, Local 21                           43424
5  SEIU - Staff and Per Diem Nurses, Local 1021        28721
6  Police Officers' Association                        24832
7  SEIU, Local 1021, RN                                24010
8  Transport Workers - Transit Operators, Local 250-A  23560
9  SEIU - Health Workers, Local 1021                   21289
10 Laborers, Local 261                                 20158

## 4. Single Vector

In [23]:
# unique(df$Job)

In [24]:
jobs.vector <- c("Chief Investment Officer","Physician Administrator, DPH","Managing Director ","Sup Psychiatric Physician Spec","Chief, Fire Department")

## 5. Five Different Lists

In [25]:
summary(df$Salaries)[c(1:3,5:6)]

     Min.   1st Qu.    Median   3rd Qu.      Max. 
-68771.78  29009.54  67899.36 100416.89 651936.71 

In [26]:
list.2016 <- list(df$Salaries[df$Year == 2016])
list.2017 <- list(df$Salaries[df$Year == 2017])
list.2018 <- list(df$Salaries[df$Year == 2018])
list.2019 <- list(df$Salaries[df$Year == 2019])
list.2020 <- list(df$Salaries[df$Year == 2020])

In [27]:
head(list.2016[[1]], n=5)
head(list.2017[[1]], n=5)
head(list.2018[[1]], n=5)
head(list.2019[[1]], n=5)
head(list.2020[[1]], n=5)

[1]  75147.32  79836.00  17806.60 136517.70  58085.17

[1]  77496.07 105997.00  97929.54 149664.90  85569.72

[1]  81096.04 115053.11 107355.32 154758.00  91988.81

[1]  91332.01 123241.00 115715.39 159394.00  65542.01

[1]  95416.01 125989.00 118150.24 162491.00  18521.00

## 6. Conditionals

In [28]:
summary(avg.salaries.df$avg_salaries)[c(1:3,5:6)]
median.salary = median(avg.salaries.df$avg_salaries)

     Min.   1st Qu.    Median   3rd Qu.      Max. 
  9774.92  58017.62  70839.40  92842.58 237005.30 

In [29]:
median.salary = median(avg.salaries.df$avg_salaries)
head(avg.salaries.df[avg.salaries.df$avg_salaries > median.salary, ])

Job.Family                    avg_salaries
1 Administrative & Mgmt (Unrep) 237005.3    
2 Unassigned                    147929.4    
3 Construction Project Mgmt     126806.2    
4 Administrative-DPW/PUC        124543.1    
5 Management                    120669.6    
6 Medical & Dental              117536.1